### PR TITLE
Fix Tag/as ProTypes

### PR DESCRIPTION
**Pull request by episodeyang**
See #2869 

#### What testing has been done on this PR?

Using function components for the tag instead.

#### What are the relevant issues?

Now, because custom components are passed in as tag, `grommet` might be adding extra styling props to those children. This leads to prop warnings as in #2870 

#### Do the grommet docs need to be updated?

No, already generating the new types.

#### Should this PR be mentioned in the release notes?

Yes.

#### Is this change backwards compatible or is it a breaking change?

it is backwards compatible.

### CODE REVIEW

File: src/js/components/Anchor/doc.js [hunk 1]
```diff
@@ -59,7 +59,9 @@ or just use children.`,
 this component. But, it can be adjusted directly via this size property, typically
 when it is not contained in a 'Heading', 'Paragraph', or 'Text'.`,
     ),
-    as: PropTypes.string.description(`The DOM tag to use for the element.`),
+    as: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
+      `The DOM tag to use for the element.`,
```
Comment (oorestisime) [inline src/js/components/Anchor/doc.js]: can we change the description to reflect this change as well?

File: src/js/components/Anchor/index.d.ts [hunk 2]
```diff
@@ -13,7 +13,7 @@ export interface AnchorProps {
   onClick?: ((...args: any[]) => any);
   reverse?: boolean;
   size?: "xsmall" | "small" | "medium" | "large" | "xlarge" | "xxlarge" | string;
-  as?: string;
+  as?: string | React.Component | React.FC | ((...args: any[]) => any);
```
Comment (oorestisime) [inline src/js/components/Anchor/index.d.ts]: i am wondering whether we should align with the typings of styled-components for this.
`keyof JSX.IntrinsicElements | React.ComponentType<any>`
both comments concern all the typings change.

File: src/js/components/Grid/doc.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
-import { describe, PropTypes } from 'react-desc';
+import {describe, PropTypes} from 'react-desc';
```
Comment (ShimiSun) [inline src/js/components/Grid/doc.js]: I think we will like to keep the spaces and be consistent across the files.

File: src/js/components/Anchor/doc.js (hunk 1 above)
Comment (oorestisime) [reply src/js/components/Anchor/doc.js]: Ok in the end this is cumbersome to do since there are different descriptions (deprecation on not) and default values

**Comment by oorestisime** [comment]
I wonder wheteher we should update the tag prop and ts definition. It is deprecated so in my understanding shouldn't get the fix (people would also move to as prop if they depend on this.)
@ShimiSun your thoughts?

File: src/js/components/Grid/doc.js [hunk 4]
```diff
@@ -64,8 +64,11 @@ space in the column axis.`,
     columns: PropTypes.oneOfType([
       PropTypes.arrayOf(
         PropTypes.oneOfType([
+          PropTypes.arrayOf(PropTypes.OneOfType([
```
Comment (oorestisime) [inline src/js/components/Grid/doc.js]: This seems unrelated with your PR. can you explain what you are doing here? The ts definitions don't need an update?

File: src/js/components/Grid/doc.js [hunk 5]
```diff
@@ -128,8 +131,11 @@ space in the row axis.`,
     rows: PropTypes.oneOfType([
       PropTypes.arrayOf(
         PropTypes.oneOfType([
+          PropTypes.arrayOf(PropTypes.OneOfType([
```
Comment (oorestisime) [inline src/js/components/Grid/doc.js]: This seems unrelated with your PR. can you explain what you are doing here? The ts definitions don't need an update?

Comment (episodeyang) [reply src/js/components/Grid/doc.js]: @oorestisime this one happened here b/c there is another unrelated bug. The prop topes for column and down is self-inconsistent. I didn't have time to file a separate issue for this and and this fix on my branch. Sorry for the confusion!!

The issue this is trying to address is the following: PropTypes of column (and row) in the past looked like 

```javascript
"xsmall"
"small"
"medium"
"large"
"xlarge"
"any CSS size"
["small", "..."]
[["small", "medium"], "..."]
```
*and* "string". "string" type is like an escape for thins that are not in the `OneOf` list.

However, the list of tuple for the lower and  higher range did not allow one to use `string` type. this is self-inconsistent. Which is why this line changes it so that the inner tuples in the ranged layout can also take non-`OneOf` strings.

File: src/js/components/Grid/doc.js (hunk 4 above)
Comment (episodeyang) [reply src/js/components/Grid/doc.js]: The `ts` definition is correct :) because it is easier to write more general type requirements than this home-brew react-docgen replacement :)

In other words, this definition is not consistent with the ts definition in its original form.

File: src/js/components/Grid/doc.js (hunk 3 above)
Comment (episodeyang) [reply src/js/components/Grid/doc.js]: I can fix this. What linter do you use? somehow my WebStorm linter does this automatically.

**Comment by episodeyang** [comment]
@oorestisime Sorry about the unrelated changes! I left the comments above, hopefully they address the issue.

You can take it over from here because you are already on this. Thanks!

In terms of the helper function, I think these changes might be good as it is. But if as is indeed shared, it makes sense to have a unified place to put it. But it would be idea if it goes into some default prototype dict instead of just a single helper function for that one PropType. 

In that case this would go beyond the scope I feel comfortable of making in the PR :) 

All in all, thanks for the quick responses!! I will try to help with more PRs in the future.

File: src/js/components/AccordionPanel/doc.js [hunk 6]
```diff
@@ -18,17 +18,17 @@ export function doc(Panel) {
...
   'accordion.icons.expand': {
     description: 'The icon to use when the panel is collapsed.',
-    type: 'React.element',
+    type: 'React.Element',
     defaultValue: '<FormDown />',
   },
   'accordion.border': {
     description: 'The border to use in the accordion.',
-    type: 'React.element',
+    type: 'React.Element',
```
Comment (episodeyang) [inline src/js/components/AccordionPanel/doc.js]: These seems to be typos. React.Element should be capitalized?

File: src/js/components/Grid/index.d.ts [hunk 7]
```diff
@@ -14,8 +14,8 @@ export interface GridProps {
   justify?: "start" | "center" | "end" | "stretch";
   justifyContent?: "start" | "center" | "end" | "between" | "around" | "stretch";
   rows?: ("xsmall" | "small" | "medium" | "large" | "xlarge" | "full" | "1/2" | "1/3" | "2/3" | "1/4" | "2/4" | "3/4" | "flex" | "auto" | "xsmall" | "small" | "medium" | "large" | "xlarge" | "full" | "1/2" | "1/3" | "2/3" | "1/4" | "2/4" | "3/4" | string | string[])[] | "xsmall" | "small" | "medium" | "large" | "xlarge" | string;
-  tag?: string | React.component | React.fc | ((...args: any[]) => any);
-  as?: string | React.component | React.fc | ((...args: any[]) => any);
+  tag?: string | React.Component | React.FC | ((...args: any[]) => any);
+  as?: string | React.Component | React.FC | ((...args: any[]) => any);
```
Comment (episodeyang) [inline src/js/components/Grid/index.d.ts]: This is an error I introduced yesterday when editing

File: src/js/components/Text/index.d.ts [hunk 8]
```diff
@@ -7,8 +7,8 @@ export interface TextProps {
   margin?: "none" | "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | {bottom?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,horizontal?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,left?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,right?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,top?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,vertical?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string} | string;
   color?: string | {dark?: string,light?: string};
   size?: "xsmall" | "small" | "medium" | "large" | "xlarge" | "xxlarge" | string;
-  tag?: string | React.component | React.fc | ((...args: any[]) => any);
-  as?: string | React.component | React.fc | ((...args: any[]) => any);
+  tag?: string | React.Component | React.FC | ((...args: any[]) => any);
+  as?: string | React.Component | React.FC | ((...args: any[]) => any);
```
Comment (episodeyang) [inline src/js/components/Text/index.d.ts]: same as above.

File: src/js/components/Grid/doc.js (hunk 3 above)
Comment (oorestisime) [reply src/js/components/Grid/doc.js]: prettier. and it actually runs as a pre-commit hook.

**Comment by oorestisime** [comment]
@episodeyang did the necessary fixes (used styled-component ts definition), rebased with current master and updated tests.
Please on your next work open a different PR for each change. Small PRs with contained changes are far more easier to review.

Thanks a lot for your contribution :)

**Comment by ShimiSun** [comment]
Looks good guys! thank you for the contribution & collaboration

**Comment by episodeyang** [comment]
Thanks guys!!
